### DISCUSSION TITLE
Issues with Lists

### DISCUSSION THREAD

**Discussion by JoelMTaylor** [retypeapp/retype#337] [Q&A] [answered]
This question is an extensions of my previous question on [Secondary and Tertiary Lists](https://github.com/retypeapp/retype/discussions/335).

I changed the format of my ordered and unordered lists to be with four spaces instead of a tab. However, that didn't fix the problem and I have run into others. 

@geoffreymcgill, you may be the only one able to view the pages due to the files and pages being restricted to only individuals in the organisation. However, I will try to include images where possible.

Firstly, I have created over 100 pages and I am now starting to get an issue that is fine in many other pages (of all the pages that have the formatting, only two have issues). These pages are [Media and Publication Consent](https://docs.ccmschools.app/edumate/enrolments/media-consent/#media-publication-consent-for-specific-purpose) and [Student Withdrawl](https://docs.ccmschools.app/edumate/enrolments/student/student/#click-guide-for-student-disenrolment-form-automatic). Both of these pages may be experiencing an issue with the formatting because the list is mixed-style (ordered and unordered together).

 1. [Media and Publication Consent](https://docs.ccmschools.app/edumate/enrolments/media-consent/#media-publication-consent-for-specific-purpose) - found in edumate\Enrolments\Media Consent.md.
    - Code: 
    ![image](https://user-images.githubusercontent.com/97919641/173269632-8b563a5a-9d7f-4777-89de-9fb07577a265.png)
    - Published page: 
    ![image](https://user-images.githubusercontent.com/97919641/173269678-3ea64771-90d0-4e20-a13f-5dbf844d3776.png)
    - Notes: just a few lines above, I have the same situation without issue.
 2. [Student Withdrawl](https://docs.ccmschools.app/edumate/enrolments/student/student/#click-guide-for-student-disenrolment-form-automatic) - found in edumate\Enrolments\Student\Student.md.
    - Code: 
    ![image](https://user-images.githubusercontent.com/97919641/173269856-4904c53a-89ce-4f0a-ba0f-721d83e54daf.png)
    - Published page: 
    ![image](https://user-images.githubusercontent.com/97919641/173269910-11b96695-0492-4d63-9c26-f47e056d3ab6.png)
    - Notes: again, there are no issues above, in a different click guide. 

I had to convert several documents from Word (```*.docx```) to markdown. I experienced issues attempting to replicate the formatting of the Word document. In particular, the Word document included many ordered and unordered lists separated by line breaks, alerts and images. The converted markdown pages can be found in the folder [Ci Anywhere](https://docs.ccmschools.app/ci-anywhere/) and their respective GitHub pages. In particular, I would draw your attention to [Disburse Purchase Card Expenses](https://docs.ccmschools.app/ci-anywhere/disburse-purchase/). As you will see from the code, the structure is a little messy with no clear and simply ordered structure. Of the nineteen steps, there is no issue until the tenth. At this point, the similar structure that has occurred earlier suddenly breaks down and it gets confused.

 3. [Disburse Purchase Card Expenses](https://docs.ccmschools.app/ci-anywhere/disburse-purchase/) - found in ci-anywhere\Disburse-purchase.md.
    - Code: 
    ![image](https://user-images.githubusercontent.com/97919641/173271026-2ed536c5-a8f2-435e-9070-9aa6c759c589.png)
    - Published page: 
    ![image](https://user-images.githubusercontent.com/97919641/173271076-ddde022c-cd89-4240-8421-6cf77e00fbd9.png)
    - Notes: I was surprised that following the structure of the indent worked. I don't know why it would break there or how to fix it (I did may tests including line breaks (```<br>```) to maintain the structure (which was moderately successful).
    - Other pages: I also experienced these issues in [Submit a Purchase Requisition](https://docs.ccmschools.app/ci-anywhere/submit/#instructions) from step 11; and [Approve Purchase Card Disbursements](https://docs.ccmschools.app/ci-anywhere/approve-purchase/) from step 4c.

I experienced some odd issues with inconsistent paragraph spacing between the primary ordered list (1,2,3). 

 4. [The Account String](https://docs.ccmschools.app/ci-anywhere/account-string/) - found at ci-anywhere\Account-String.md
    - Code: each new section of the ordered list has the same consistent formatting ![image](https://user-images.githubusercontent.com/97919641/173271904-6c47a085-eb27-4288-be84-7dfd0f272d6e.png)
    - Published page: but the eighth point has a smaller/no paragraph space above (between 7d and 8) ![image](https://user-images.githubusercontent.com/97919641/173272089-c8eeea20-5549-4ca4-bddd-0331b23178ac.png)
    - Other pages: [Approve a Purchase Requisition](https://docs.ccmschools.app/ci-anywhere/approve/) has small/no paragraph spacing above (spacing is present due to the image in line); [Approve Purchase Card Disbursements](https://docs.ccmschools.app/ci-anywhere/approve-purchase/) appears to start off with spacing before it stops from step 3; [Disburse Purchase Card Expenses](https://docs.ccmschools.app/ci-anywhere/disburse-purchase/) maintains consistent large paragraph spacing between each primary ordered list, but has a hiccup between 5d and 5e (probably due to a kind of line break - which is supposed to maintain the same in indentation but without a ordered number at the start); [Submit a Purchase Requisition](https://docs.ccmschools.app/ci-anywhere/submit/#instructions) from step 13 reverts to small/no paragraph spacing; and [Transmit or Amend a Purchase Order](https://docs.ccmschools.app/ci-anywhere/transmit/#instructions) from step 2.

Similar to maintaining the ordered structure (point three), I had issues inserting an image at the same indent of a sub-point and maintaining the same indent or structure as required. 
 5. [Approve a Purchase Requisition](https://docs.ccmschools.app/ci-anywhere/approve/) - found at ci-anywhere\Approve.md.
    - Code: 
    ![image](https://user-images.githubusercontent.com/97919641/173274447-43ff9949-f61d-4cc3-a784-b4bda0db1129.png)
    - Published page: 
    ![image](https://user-images.githubusercontent.com/97919641/173274484-ef781aa1-9ada-4b45-9b28-11efa3b7b234.png)
    - Notes: this occurs for any image at a second or tertiary level (not at the primary/no level) and that I am trying to achieve the same as is occurring with the images after 'code' or 'published pages'.
    - Other pages: [Receive Goods or Service from a Purchase Order](https://docs.ccmschools.app/ci-anywhere/receive/); and [Transmit or Amend a Purchase Order](https://docs.ccmschools.app/ci-anywhere/transmit/#instructions).

That is a lot of information. I know that Markdown is going to be different to Word and where formatting cannot be completed the same I will just have to life with or work around. However, I just want to ask and check first - because there have been many times where I was unaware of an issue or solution (such as the [duplicate document 'idea'](https://github.com/retypeapp/retype/discussions/334)).

Thank you again for all your assistance.

**Comment by geoffreymcgill** [comment]
Hi @JoelMTaylor, 

I tried the source for [Disburse-purchase.md](https://github.com/ccmltd/docs-corportatesystems/blob/main/ci-anywhere/Disburse-purchase.md?plain=1) in the [CommonMark playground](https://spec.commonmark.org/dingus/) and the formatting began failing around item 10 as well. I noticed that you had an extra space in front of each list item. Once I remove that extra space char, the formatting appears to work as expected.

<img width="412" alt="Screen Shot 2022-06-13 at 6 52 03 PM" src="https://user-images.githubusercontent.com/62210/173471125-1bc657ee-33d5-4beb-86a3-c49c1d71be0b.png">

I'll have to read through the CommonMark [spec](https://spec.commonmark.org/0.28/#lists) to figure out the exactly why a single digit list marker with a leading space works, but a double-digit list marker with a leading space fails.

Removing those extra leading spaces should solve the formatting issues. I will do some research too.

**Reply by JoelMTaylor** [reply]
> I noticed that you had an extra space in front of each list item. Once I remove that extra space char, the formatting appears to work as expected.

Thanks @geoffreymcgill for the advice. Removing the space at the front of the ordered list at the primary level fixed the issues with (1), (2) and (3).

**Comment by JoelMTaylor** [comment]
I have been playing around again, trying to find a solution that appears fine. I tried to test things with the CommonMark Playground, but found that it did not display things consistently to Retype (some for better, others for worse).

Item 4 was of particular annoyance to me since I am inclined to noticing and having issue with even very minor inconsistencies. I tried various solutions, but each required some use of ```<br>``` in order to hack a solution. 

Before I explain what I have tested, I want to mention that I have defined a few terms at the bottom of this post to ensure that nothing of this explanation is lost. Review this if you wish. If I have any of these terms incorrect, please let me know.

The consistent pattern of ordered, unordered and mixed lists seems to be one with no paragraph spacing between individual items. That is, there is no gap between 1,2,3 or 1,a,b,2. Each appears close. For example:
![image](https://user-images.githubusercontent.com/97919641/173981845-64f41795-9979-4948-83fe-ec1b415fc3cd.png)

However, this could be changed by entering one extra line break in one location. By doing so, each list item has paragraph spacing.
![image](https://user-images.githubusercontent.com/97919641/173982041-1fb5952f-d665-4569-b4ef-8bdce8b7977c.png)

While I don't know *why* this happens (I haven't read through the CommonMark documentation enough), this seems to be a pleasing option and could be used as required. In my directionless testing, I can confirm this to be true for me - in the most part. There is one issue which seems to be at the root of my question above (point three).

It seems to me that there is an issue with how the paragraph spacing is created when an additional line break is added at a list item (example two just above). In the code excerpt below, one would presume that the paragraph spacing would mimic that of the line breaks. However, this is not what occurs.

- page: [Account String](https://docs.ccmschools.app/ci-anywhere/account-string/) - found at ci-anywhere\Account-String.md
- code:
![image](https://user-images.githubusercontent.com/97919641/173983493-8d6f843f-7b93-4043-adcb-8e6df4782863.png)
- published page:
![image](https://user-images.githubusercontent.com/97919641/173983527-4dde9f41-ac19-40cc-aa1f-876ed669ddfa.png)

I consistently find that when a primary ordered list item (e.g. 1,2,3; I have not tested with unordered lists) is preceded by secondary ordered list item (e.g. a,b,c), that primary item fails to have the correct paragraph spacing. The only solution that I could hack was to have two breaks in order to force the appearance of the correct paragraph spacing (one break would not be sufficient). As you can probably guess, extra line breaks was not successful (only two breaks). 

If creating lists with a small amount of paragraph spacing is built in to CommonMark (according to example two), then I _suppose_ this is an issue to be fixed. 

Similarly, I noticed a behaviour with alerts within a list that did the opposite - added paragraph spacing after when none was expected. When using on single break spacing (tight list like example one), an alert adds the paragraph space succeeding itself, regardless of the context (that is, indent level).

- page: [Approve a Purchase Requisition](https://docs.ccmschools.app/ci-anywhere/approve/) - found at ci-anywhere\Approve.md.
- code:
![image](https://user-images.githubusercontent.com/97919641/173984621-dd04e5a8-b89a-4564-b860-30b9eb2a6639.png)
- published page:
![image](https://user-images.githubusercontent.com/97919641/173984676-cd03f8f1-991d-4b73-ac21-ffc1b5a8cbfe.png)

I suppose this also to be an issue. Also, one may notice that tabs are used instead of spaces for indents. I regularly switched between the two to make sure one or other was not the issue. I have not found an instance where spaces work and tabs do not. I would further add that I was not able to hack a fix for this as adding a break ```<br>``` at the end of the final ```!!!``` breaks (damages) the code. 

So, while I was able to find a hack for displaying the list with extra paragraph spacing, I preferred the more consistent (to my other pages) single spacing. This provided me another issue to solve - point four of my main post above. The obvious answer would be to have an extra line break between the primary (or secondary) item and the image. However, doing so causes the paragraph spacing issue I am also trying to avoid. The hack that I found was to insert a break ```<br>``` at the end of the preceding line. This would have to be done for images and ```<mark>``` (edit: it is with images and indented text [without leading number/letter/symbol]). I initially thought that it would need to be done the line preceding and proceeding (at the end of the image line), but the later turned out to unnecessary. 

- page: [Receive Goods or Service from a Purchase Order](https://docs.ccmschools.app/ci-anywhere/receive/) - found at ci-anywhere\Receive.md.
- code:
![image](https://user-images.githubusercontent.com/97919641/173986619-13219e09-1a34-4b5d-b557-3ea9b661c8de.png)
- published page:
![image](https://user-images.githubusercontent.com/97919641/173986707-4c52eb4d-4a88-40fa-81a3-7b58340fdfc2.png)

@geoffreymcgill, if you have any solutions or ideas, or you perceive any of these to be issues that you desire to fix (at any time in the future), please let me know.

While it would be *nice* to not require any hacks in order to achieve a desired appearance, it is plain to understand that Markdown is designed for simplicity and to be light and one cannot have it both ways. If a solution is to be found to added in the future, that would be great. Otherwise, we make it work or make do. 
___
Definitions of terms:
1. A "break" refers to the use of ```<br>```. Normally, this is done at the end of one line instead of entering two "line breaks".
2. A "line break" refers to pressing enter at the end of a line to create a new line below. Normally, to separate sections of text that are not linked together (distinct "paragraphs"), two line breaks (enter is pressed twice) to create a gab between the two sections. Line breaks do not have a symbol in Studio Code, but just increase the numbering of the line by one.
3. A "paragraph" refers to two sections of text that are separated in order to show their different (that is, not consistent or continuous) nature. 
4. A "paragraph spacing" refers to the space between two paragraphs that visually demonstrate their distinction.

**Comment by geoffreymcgill** [comment, marked as answer]
Hi @JoelMTaylor, 

Regarding the lists and spacing, the following thread discussing CommonMark Loose Lists vs Tight Lists should help provide some insight, see https://github.com/retypeapp/retype/issues/67

Instead of adding `<br>` line breaks, in Markdown you can end a line with a `\` to force a hard line break. The `\` gets converted to a `<br />` in the final generated HTML.

As well, the [`lineBreaks`](https://retype.com/configuration/project/#linebreaks) config might also be something you would like to experiment with. Try the following setting in your `retype.yml`:

```yml
markdown:
  lineBreaks: hard
```

Hope this helps.

**Reply by JoelMTaylor** [reply]
Hi @geoffreymcgill, 

That is very useful - the links are most helpful. [[```lineBreaks```](https://retype.com/configuration/project/#linebreaks) has the wrong hyperlink - but it was easy to find.]

It is particularly interesting about [Andrey Chechel's solution](https://github.com/retypeapp/retype/issues/67#issuecomment-878198101) to creating a loose list after a single item: ```[]()```.

Do you have any ideas for creating the correct spacing following an alert? I tried using a ```\```, but it behaves similarly to a ```<br>```. [A reminder of the situation below.]
![image](https://user-images.githubusercontent.com/97919641/174148646-59267308-0931-4527-a760-b1f67df89e6c.png)
